src/BalanceTx.purs

```diff
@@ -382,18 +385,29 @@ balanceTx unattachedTx@(UnattachedUnbalancedTx { unbalancedTx: t }) = do
       note (GetWalletAddressError' CouldNotGetNamiWalletAddress)
     utxos <- ExceptT $ utxosAt ownAddr <#>
       (note (UtxosAtError' CouldNotGetUtxos) >>> map unwrap)
+    collateral <- ExceptT $ getWalletCollateral <#>
+      note (GetWalletCollateralError' CouldNotGetNamiCollateral)
+    wallet <- asks _.wallet
     let
       -- Combines utxos at the user address and those from any scripts
       -- involved with the contract in the unbalanced transaction.
       allUtxos :: Utxo
       allUtxos = utxos `Map.union` utxoIndexToUtxo networkId utxoIndex
 
+      -- After adding collateral, we need to balance the inputs and
+      -- non-Ada outputs before looping, i.e. we need to add input fees
+      -- for the Ada only collateral. No MinUtxos required. In fact perhaps
+      -- this step can be skipped and we can go straight to prebalancer.
+      unbalancedCollTx = case wallet of
+        Just (Nami _) -> addTxCollateral unbalancedTx' collateral
```

Comment:
We need to add the collateral from Gero, no? Is this what you were referring to in your top-level comment?